content/posts/i-like-rust.md

```diff
@@ -0,0 +1,218 @@
+In this case it's possible to `panic`. When a program panics it will crash by unwinding its call-stack, optionally printing a stack trace.
+You won't be able to catch panics.
+So Rust's approach is simple and pragmatic. A fault condition should be either recoverable or not.
+
+## Error messages
+
+Error messages in Rust are awesome. They're super helpful, concise and correct. As the compiler itself is very strict in general you will encounter compilation errors where you don't know how to resolve the issue and you have to rethink your approach. But the error messages are as helpful as they could possible get. An unspecific or unclear error message is considered a bug by the Rust team, and you should report it if you ever encounter one. Some compilers give vague error messages and some even point to the wrong source files! This for example happens when working with templates in C++. Working with such error messages is a nightmare.
+
+Runtime errors in Rust are a rarer phenomenon but they will be just as clear as the error messages by the compiler. Most of the time the concise messages are enough to track down the mistake. Optionally backtraces can be enabled as Rust provides a tiny runtime.
+
+If you create a stack overflow with C++ the error message could look like this[^6]:
+
+```
+HelloWorld.exe (process 15916) exited with code -1073741571.
+```
+
+Or it might look like this:
+
+```
+Segmentation fault
+```
+
+With Rust the error message looks as follows:
+
+```
+thread 'main' has overflowed its stack
+fatal runtime error: stack overflow
+```
+
+With Rust you don't have to be a coding wizard to understand error messages. In my attempt to create a stack overflow in Rust, the compiler also warned me about the fact that my `function cannot return without recursing`.
+
+## Dependencies
+
+When writing software these days, you will probably sooner or later depend on external libraries. This is why dependency management shouldn't be a pain. Many languages evolved before streamlined dependency management was a thing. I think NPM, the package manager for NodeJS was the first good dependency management system for a programming language. I became very popular and might be the main reason why NodeJS itself got so popular. One thing I dislike about the NPM ecosystem is that dependency graphs tend to be huge. This is not a problem of NPM itself. To some extent it might be a problem of NodeJS or JavaScript. Simple things like cloning objects or comparing objects for equality are not possible out of the box. When people in an ecosystem tend to add dependencies to their libraries without much thought it becomes the norm. Rustaceans seem to be more conservative about adding dependencies to their projects. Rust's build system and package manager, Cargo, is very similar to NPM with a few improvements. With NPM, for example, I regularly forget to install or update the dependencies with `npm install`, whereas Cargo does this automatically for you.
```

Comment:
> Rustaceans seem to be more conservative about adding dependencies to their projects.

Huh, I've seen a similar attitude towards adding dependencies as in Node...
At least, it's a bold statement, which I'd back up with data or a reference.